lib/galaxy/tools/actions/__init__.py

```diff
@@ -193,7 +193,22 @@ def _collect_inputs(self, tool, trans, incoming, history, current_user_roles):
         # Collect any input datasets from the incoming parameters
         inp_data = self._collect_input_datasets(tool, incoming, trans, history=history, current_user_roles=current_user_roles)
 
-        return history, inp_data, inp_dataset_collections
+        # grap tags from incoming HDAs
+        preserved_tags = {}
+        for data in inp_data.values():
+            for tag in [t for t in data.tags if t.user_tname == 'name']:
+                preserved_tags[tag.value] = tag
+
+        # grap tags from incoming HDCAs
+        for key, collection_pairs in inp_dataset_collections.items():
```

Comment:
given that we're not using `key` we could just use `for collection_pairs in inp_dataset_collections.values():`